components/higher-order/if-condition/index.js

```diff
@@ -0,0 +1,37 @@
+/**
+ * WordPress dependencies
+ */
+import { getWrapperDisplayName } from '@wordpress/element';
+
+/**
+ * Higher-order component creator, creating a new component which renders if
+ * the given condition is satisfied or with the given optional prop name.
+ *
+ * @param {Function} predicate Function to test condition.
+ * @param {?string}  propName  Optional name of prop passed to component with
+ *                             result of predicate. If provided, component will
+ *                             always render even if predicate returns false.
+ *
+ * @return {Function} Higher-order component.
+ */
+const ifCondition = ( predicate, propName ) => ( WrappedComponent ) => {
+	const EnhancedComponent = ( props ) => {
+		const isOk = predicate( props );
+
+		if ( ! propName && ! isOk ) {
+			return null;
+		}
+
+		if ( propName ) {
+			props = { ...props, [ propName ]: isOk };
+		}
+
+		return <WrappedComponent { ...props } />;
+	};
+
+	EnhancedComponent.displayName = getWrapperDisplayName( WrappedComponent, 'condition' );
```

Comment:
Should it be getWrapperDisplayName( WrappedComponent, 'ifCondition' );?

Reply:
We omit prefixes in other places, but we might change this if this is confusing.

Reply:
>We omit prefixes in other places, but we might change this if this is confusing.

I'm thinking particularly with https://github.com/WordPress/gutenberg/pull/5206#discussion_r170247231 in mind, we might want to do this, since we could have two higher-order components with similar names distinguished on prefix, e.g. `withViewportMatch`, `ifViewportMatches`.

Also to the point of https://github.com/WordPress/gutenberg/pull/5206#discussion_r170274728, I'm starting to think maybe we could use a new higher-order component utility in place of the current `getWrapperDisplayName` to allow for easier HOC composition.

For example:

```js
export default createHigherOrderComponent(
	'ifViewportMatches',
	( query ) => compose( [
		withViewportMatch( query ),
		ifCondition( ( props ) => props.isViewportMatch ),
	] )
);
```

Reply:
@aduth, this is actually a great idea even for the regular HOCs. See:

```js
const ifCondition = ( predicate ) => ( WrappedComponent ) => {
	const EnhancedComponent = ( props ) => {
		if ( ! predicate( props ) ) {
			return null;
		}

		return <WrappedComponent { ...props } />;
	};

	EnhancedComponent.displayName = getWrapperDisplayName( WrappedComponent, 'ifCondition' );

	return EnhancedComponent;
};
```
vs.
```js
const ifCondition = createHigherOrderComponent(
	'ifCondition',
	( predicate ) => ( WrappedComponent ) => {
		return ( props ) => {
			if ( ! predicate( props ) ) {
				return null;
			}

			return <WrappedComponent { ...props } />;
		};
	},
);
```

I'm voting to refactor all HOCs to use the helper you proposed and update all prefixes to match what has been discussed here.

Reply:
Implementation wise it might be non-trivial task because some HOC take additional params and requires a higher-order function to be involved :)

Reply:
>Implementation wise it might be non-trivial task because some HOCs take additional params and require a higher-order function to be involved :)

Yeah, I was thinking about this too. One option is to test whether the argument passed is a class extending `Component`, but this would mean that all higher-order components would have to be implemented as the full-on class form. Maybe not a big deal to impose this?